[parametrization] fix `requires_grad` propagation

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124888

Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[parametrization] fix `requires_grad` propagation and docs

Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -418,8 +420,8 @@ def register_parametrization(
 
-    This method is called on the unparametrized tensor when the first parametrization
-    is registered to compute the initial value of the original tensor.
+    This method is called on the unparametrized tensor (original ``module.weight``) when the first paramet`rization
```
```suggestion
    This method is called on the unparametrized tensor (original ``module.weight``) when the first parametrization
```

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -432,6 +434,17 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
     with names ``original0``, ``original1``,...
 
+
+    What happens during eager mode `forward` run:
```
```suggestion
    What happens during `forward` run:
```

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -432,6 +434,17 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
 
+
+    What happens during eager mode `forward` run:
+        `parametrizations` are implemented with `ModuleList`, so each `parametrization` will be a `Module`
```
I'm not sure about this description? 
As discussed above, parametrizations field is actually a a ModuleDict of ParametrizationList

jerryzh168 replied:
oh sorry, let me update this

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lezcano commented on test/nn/test_parametrization.py:
```diff
@@ -1414,6 +1414,26 @@ def fn(weight):
+        # One parametrization with unsafe=True
+        parametrize.register_parametrization(model, "weight", SplitAndCat())
+        # making sure the decomposed Tensors both have requires_grad == False
+        self.assertFalse(model.parametrizations.weight.original0.requires_grad)
+        self.assertFalse(model.parametrizations.weight.original1.requires_grad)
```
Probably also check that
```suggestion
        self.assertFalse(model.weight.requires_grad)
```

lezcano replied:
And check that this is respected when you assign to it?

lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -432,6 +435,18 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
 
+
+    What happens during `forward` run:
+        `parametrizations` are implemented with `ModuleDict` mapping from tensor_name to a
```
Use double backticks

lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -432,6 +435,18 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
+        `parametrizations` are implemented with `ModuleDict` mapping from tensor_name to a
+        `ParametrizationList` (a subclass of `ModuleList`), so each `parametrization` will be a `ModuleList`
+        when the model calls `self.weight` in forward, instead of returnning the original weight, it
+        will return the `parametrization`s (`ModuleList`) that's registered for the weight (for the first
```
```suggestion
        will return the `parametrization`\ s (`ModuleList`) that's registered for the weight (for the first
```

lezcano reviewed (commented):
Actually, it would be better, rather than spelling out what the parametrisation does in an already long doc, adding a link to the tutorial:
https://pytorch.org/tutorials/intermediate/parametrizations.html

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> Actually, it would be better, rather than spelling out what the parametrisation does in an already long doc, adding a link to the tutorial: [pytorch.org/tutorials/intermediate/parametrizations.html](https://pytorch.org/tutorials/intermediate/parametrizations.html)

I read the tutorial and doesn't seem to help understanding what we are doing actually.. especially what right_inverse is doing etc.

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lezcano commented:
The `right_inverse` is discussed at https://pytorch.org/tutorials/intermediate/parametrizations.html#initializing-parametrizations

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> The `right_inverse` is discussed at [pytorch.org/tutorials/intermediate/parametrizations.html#initializing-parametrizations](https://pytorch.org/tutorials/intermediate/parametrizations.html#initializing-parametrizations)

right, I guess main thing is what happens during registration and during forward seems to be missing, and also explaining what it is doing (how it is implemented) that could help understanding as well. but it may be just me

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation and docs"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors
Also added some docs to explain what happens during registration and during forward

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lezcano commented:
What happens in the forward is explained via examples at https://pytorch.org/tutorials/intermediate/parametrizations.html#concatenating-parametrizations. 

We also say in the docs:
>Parametrizations may be concatenated by registering several parametrizations on the same attribute.

If it's not clear enough, we should probably extend that section of the tutorial.

jerryzh168 commented:
> What happens in the forward is explained via examples at [pytorch.org/tutorials/intermediate/parametrizations.html#concatenating-parametrizations](https://pytorch.org/tutorials/intermediate/parametrizations.html#concatenating-parametrizations).
> 
> We also say in the docs:
> 
> > Parametrizations may be concatenated by registering several parametrizations on the same attribute.
> 
> If it's not clear enough, we should probably extend that section of the tutorial.

yeah it doesn't explain how this relates to right_inverse I think

jerryzh168 commented:
@lezcano I want to land this PR first, let me know if your comments are blocking or not, I can also just revert the changes to the docs

lezcano reviewed (approved):
Could we split the functionality and the comments into two PRs or just revert the comments?

The tests and functionality LGTM tho. Approving to unblock that part.

jerryzh168 committed:
Update on "[parametrization] fix `requires_grad` propagation"


Summary:
Previously the `requires_grad` is not propagated from original Tensor to decomposed tensors

Test Plan:
python test/test_parametrization.py -k test_register_parametrization_no_grad

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge
